core/src/main/java/org/apache/iceberg/avro/ValueReaders.java

```diff
@@ -228,50 +228,112 @@ public static List<Pair<Integer, ValueReader<?>>> buildReadPlan(
       Map<Integer, ?> idToConstant,
       BiFunction<Type, Object, Object> convert) {
     Map<Integer, Integer> idToPos = idToPos(expected);
-
     List<Pair<Integer, ValueReader<?>>> readPlan = Lists.newArrayList();
-    List<Schema.Field> fileFields = record.getFields();
-    for (int pos = 0; pos < fileFields.size(); pos += 1) {
+    addFileFieldReadersToPlan(readPlan, record.getFields(), fieldReaders, idToPos, idToConstant);
+    addMissingFileReadersToPlan(readPlan, idToPos, expected, idToConstant, convert);
+    return readPlan;
```

Comment:
I didn't end up refactoring like we do in `ParquetValueReaders`, it ended up being a bit hard to read  considering avro positions compared to just splitting into two separate methods, one where we just add the file field readers to the read plan first and then we add the missing readers to the plan. That also eliminates cyclomatic complexity